### PR TITLE
[clang][OpenCL][CodeGen][AMDGPU] Do not use `private` as the default AS for when `generic` is available

### PR DESCRIPTION
**Pull request by AlexVlx**
Currently, for AMDGPU, when compiling for OpenCL, we unconditionally use `private` as the default address space. This is wrong for cases where the `generic` address space is available, and is corrected via this patch. In general, this AS map abuse is a bad hack and we should re-work it altogether, but at least after this patch we will stop being incorrect for e.g. OpenCL 2.0.

### CODE REVIEW

File: clang/test/CodeGenOpenCL/addr-space-struct-arg.cl [hunk 1]
```diff
@@ -1,9 +1,10 @@
-// RUN: %clang_cc1 %s -emit-llvm -o - -O0 -ffake-address-space-map -triple i686-pc-darwin | FileCheck -enable-var-scope -check-prefixes=ALL,X86 %s
-// RUN: %clang_cc1 %s -emit-llvm -o - -O0 -triple amdgcn | FileCheck -enable-var-scope -check-prefixes=ALL,AMDGCN %s
-// RUN: %clang_cc1 %s -emit-llvm -o - -cl-std=CL2.0 -O0 -triple amdgcn | FileCheck -enable-var-scope -check-prefixes=ALL,AMDGCN,AMDGCN20 %s
-// RUN: %clang_cc1 %s -emit-llvm -o - -cl-std=CL1.2 -O0 -triple spir-unknown-unknown-unknown | FileCheck -enable-var-scope -check-prefixes=SPIR %s
-// RUN: %clang_cc1 %s -emit-llvm -o - -cl-std=CL3.0 -O0 -triple amdgcn -cl-ext=+__opencl_c_program_scope_global_variables | FileCheck -enable-var-scope -check-prefixes=ALL,AMDGCN,AMDGCN20 %s
-// RUN: %clang_cc1 %s -emit-llvm -o - -cl-std=CL3.0 -O0 -triple amdgcn | FileCheck -enable-var-scope -check-prefixes=ALL,AMDGCN %s
+// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py UTC_ARGS: --check-attributes --check-globals all --include-generated-funcs --version 5
```
Comment (arsenm) [inline clang/test/CodeGenOpenCL/addr-space-struct-arg.cl:1]: Separately commit conversions of tests to generated checks

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 2]
```diff
@@ -5903,7 +5904,7 @@ RValue CodeGenFunction::EmitBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
       auto Call = RValue::get(
           EmitRuntimeCall(CGM.CreateRuntimeFunction(FTy, Name), Args));
       if (TmpSize)
-        EmitLifetimeEnd(TmpSize, TmpPtr);
+        EmitLifetimeEnd(TmpSize, TmpPtr->stripPointerCasts());
```
Comment (arsenm) [inline clang/lib/CodeGen/CGBuiltin.cpp:5907]: I'm surprised CreateMemTemp isn't returning the raw alloca. I would expect it to return the direct reference, and not the addrspacecasted one. 

Assuming fixing that is hard, should just do the strip once and use it consistently

File: clang/lib/CodeGen/CGBlocks.cpp [hunk 3]
```diff
@@ -1397,7 +1397,8 @@ void CodeGenFunction::setBlockContextParameter(const ImplicitParamDecl *D,
       DI->setLocation(D->getLocation());
       DI->EmitDeclareOfBlockLiteralArgVariable(
           *BlockInfo, D->getName(), argNum,
-          cast<llvm::AllocaInst>(alloc.getPointer()), Builder);
+          cast<llvm::AllocaInst>(alloc.getPointer()->stripPointerCasts()),
```
Comment (arsenm) [inline clang/lib/CodeGen/CGBlocks.cpp:1400]: I would hope alloc would be the direct alloca reference to begin with

File: clang/test/CodeGenOpenCL/amdgcn-automatic-variable.cl [hunk 4]
```diff
@@ -1,67 +1,111 @@
-// RUN: %clang_cc1 -O0 -cl-std=CL1.2 -triple amdgcn---amdgizcl -emit-llvm %s -o - | FileCheck -check-prefixes=CHECK,CL12 %s
-// RUN: %clang_cc1 -O0 -cl-std=CL2.0 -triple amdgcn---amdgizcl -emit-llvm %s -o - | FileCheck -check-prefixes=CHECK,CL20 %s
+// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py UTC_ARGS: --version 5
+// RUN: %clang_cc1 -O0 -cl-std=CL1.2 -triple amdgcn---amdgizcl -emit-llvm %s -o - | FileCheck -check-prefixes=CL12 %s
```
Comment (arsenm) [inline clang/test/CodeGenOpenCL/amdgcn-automatic-variable.cl:2]: Convert tests in separate commit

Also we should get rid of amdgizcl references, that's a leftover from the address space map conversion

File: clang/lib/CodeGen/CGBuiltin.cpp (hunk 2 above)
Comment (AlexVlx) [reply clang/lib/CodeGen/CGBuiltin.cpp:5907]: I would extremely strongly prefer not to mess with this ball of yarn or other pieces of functionality (specifically, `CreateMemTemp`), as part of what is essentially a point fix of a long standing bug. I'll leave a TODO in place for someone that is braver and more skilled.

Comment (arsenm) [reply clang/lib/CodeGen/CGBuiltin.cpp:5907]: Can still just do the stripPointerCasts once above

**Review by arsenm — APPROVED** [review]

File: clang/test/CodeGenOpenCL/addr-space-struct-arg.cl (hunk 1 above)
Comment (AlexVlx) [reply clang/test/CodeGenOpenCL/addr-space-struct-arg.cl:1]: Done.

File: clang/test/CodeGenOpenCL/amdgcn-automatic-variable.cl (hunk 4 above)
Comment (AlexVlx) [reply clang/test/CodeGenOpenCL/amdgcn-automatic-variable.cl]: Done.

**Comment by AlexVlx** [comment]
> LLVM Buildbot has detected a new failure on builder `openmp-offload-amdgpu-runtime` running on `omp-vega20-0` while building `clang` at step 7 "Add check check-offload".
> 
> Full details are available at: https://lab.llvm.org/buildbot/#/builders/30/builds/8694
> 
> Here is the relevant piece of the build log for the reference

@jhuber6 apologies for the ping but any idea why a pure OCL change would trigger a failure in our OMP offload RT?

**Comment by jhuber6** [comment]
> > LLVM Buildbot has detected a new failure on builder `openmp-offload-amdgpu-runtime` running on `omp-vega20-0` while building `clang` at step 7 "Add check check-offload".
> > Full details are available at: https://lab.llvm.org/buildbot/#/builders/30/builds/8694
> > Here is the relevant piece of the build log for the reference
> 
> @jhuber6 apologies for the ping but any idea why a pure OCL change would trigger a failure in our OMP offload RT?

Seems to be green now, but the tests in general are flaky. I don't put much stock in random OpenMP failures anymore.

**Pull request merged**